gu-client/src/async.rs

```diff
@@ -0,0 +1,297 @@
+                        ),
+                        _ => future::Either::B(future::err(Error::InternalError)),
+                    }).and_then(|answer_json: PeerInfo| {
+                        future::ok(Peer {
+                            peer_info: answer_json,
+                            hub_session: hub_session_copy,
+                        })
+                    }),
+            ),
+            _ => future::Either::B(future::err(Error::CannotCreateRequest)),
+        };
+    }
+}
+
+/// Large binary object.
+#[derive(Debug)]
+pub struct Blob {
+    hub_session: HubSession,
+    blob_id: String,
+}
+
+struct StreamError<T: std::error::Error> {
+    inner: T,
+}
+
+impl Blob {
+    /// uploads blob represented by a stream
+    pub fn upload_from_stream<S, T>(&self, stream: S) -> impl Future<Item = (), Error = Error>
+    where
+        S: Stream<Item = Bytes, Error = T> + 'static,
+        T: Into<actix_web::Error>,
+    {
+        let url = format!(
+            "{}sessions/{}/blob/{}",
+            self.hub_session.driver.driver_inner.url, self.hub_session.session_id, self.blob_id
+        );
+        let request = match client::ClientRequest::post(url).streaming(stream) {
+            Ok(r) => r,
+            _ => return future::Either::A(future::err(Error::CannotCreateRequest)),
+        };
+        future::Either::B(
+            request
+                .send()
+                .map_err(|_| Error::CannotSendRequest)
+                .and_then(|response| match response.status().as_u16() {
+                    200 => future::ok(()),
+                    _ => future::err(Error::InternalError),
+                }),
+        )
+    }
+}
+
+/// Peer node.
+#[derive(Clone)]
+pub struct Peer {
+    hub_session: HubSession,
+    pub peer_info: PeerInfo,
+}
+
+impl Peer {
+    /// creates new peer session
+    /// assumption: url is /sessions/{session_id}/peer/{peer_id}/peer_sessions (was: /hd)
+    pub fn new_session(
+        &self,
+        builder: PeerSessionInfoBuilder,
+    ) -> impl Future<Item = PeerSession, Error = Error> {
+        let url = format!(
+            "{}sessions/{}/peer/{}/peer_sessions",
+            self.hub_session.driver.driver_inner.url,
+            self.hub_session.session_id,
+            self.peer_info.node_id.to_string()
+        );
+        let session_info = match builder.build() {
+            Ok(r) => r,
+            _ => return future::Either::A(future::err(Error::InvalidPeerSessionParameters)),
+        };
+        let request = match client::ClientRequest::post(url).json(session_info) {
+            Ok(r) => r,
+            _ => return future::Either::A(future::err(Error::CannotCreateRequest)),
+        };
+        let peer_copy = self.clone();
+        future::Either::B(
+            request
+                .send()
+                .map_err(|_| Error::CannotSendRequest)
+                .and_then(|response| {
+                    if response.status().as_u16() != 201 {
```

Comment:
You may want to use https://docs.rs/http/0.1.1/http/status/struct.StatusCode.html instead of magic constants.

Reply:
You are right. Fixed.